stdlib/public/SDK/Foundation/DataThunks.m

```diff
@@ -91,8 +91,46 @@ static NSInteger _NSWriteToFileDescriptor(int32_t fd, const void *buffer, NSUInt
     return length - numBytesRemaining;
 }
 
-extern NSError *_NSErrorWithFilePath(NSInteger code, id pathOrURL);
-extern NSError *_NSErrorWithFilePathAndErrno(NSInteger posixErrno, id pathOrURL, BOOL reading);
+static NSError *_NSErrorWithFilePath(NSInteger code, id pathOrURL) {
+    NSString *key = [pathOrURL isKindOfClass:[NSURL self]] ? NSURLErrorKey : NSFilePathErrorKey;
+    return [NSError errorWithDomain:NSCocoaErrorDomain code:code userInfo:[NSDictionary dictionaryWithObjectsAndKeys:pathOrURL, key, nil]];
+}
+
+static NSError *_NSErrorWithFilePathAndErrno(NSInteger posixErrno, id pathOrURL, BOOL reading) {
+    NSInteger code;
+    if (reading) {
+        switch (posixErrno) {
+            case EFBIG:         code = NSFileReadTooLargeError; break;
+            case ENOENT:	code = NSFileReadNoSuchFileError; break;
+            case EPERM:
+            case EACCES:	code = NSFileReadNoPermissionError; break;
+            case ENAMETOOLONG:  code = NSFileReadInvalidFileNameError; break;
+            default:		code = NSFileReadUnknownError; break;
+        }
+    } else {
+        switch (posixErrno) {
```

Comment:
Would be nice to extract this switch into a `__attribute__((__always_inline__))` function that can be re-used here.  Something like `NSErrorFromPOSIXError`